.travis.yml

```diff
@@ -11,95 +12,131 @@ git:
+# installations in sequence etc.
+install:
+  # Install chartpress, pytest, etc.
+  - pip3 install --no-cache-dir -r dev-requirements.txt
+
+  # Fix to avoid rebuilding images already built
+  - remove_docker_mirror_on_travis
+
+  # Setup our k3s based Kubernetes cluster and use its kubectl as well
+  - setup_k3s
   - setup_helm
-  - ./ci/travis-docker-fix
-  - ./dev kind create
   - helm repo add jupyterhub https://jupyterhub.github.io/helm-chart/
   - helm repo update
-  - ./dev upgrade --chart jupyterhub/jupyterhub --version 0.8.2
-  - ./dev upgrade --chart ./jupyterhub
-  - ./dev port-forward
-  - ./dev test
+
+  # Install a local ACME server
+  - helm install pebble jupyterhub/pebble --values dev-config.yaml
+
+  # Build our images if needed and update values.yaml with the tags
+  - chartpress
+
+  # Pre-install JupyterHub
+  - await_dns
+  - await_pebble
+script:
+  # Install JupyterHub
+  - helm install jupyterhub ./jupyterhub --values dev-config.yaml
+  - await_jupyterhub
+  - await_tls_cert_acquisition
+
+  # Run tests
+  - pytest --verbose --exitfirst ./tests
+
 env:
-  ## NOTE: The environment variables will be expanded to multiple jobs. For
-  ##       additional individual jobs, only the first entry is used.
-  ##
-  ## ref: https://docs.travis-ci.com/user/build-stages/#build-stages-and-build-matrix-expansion
-  ##
-  ##
-  ## Z2JH_KUBE_VERSION should match a released kindest/node image tag, but they
-  ## are currently not automatically published.
-  ##
-  ## ref: https://hub.docker.com/r/kindest/node/tags
-  ## ref: https://github.com/kubernetes-sigs/kind/issues/197
-  ##
-  - Z2JH_KUBE_VERSION=1.13.12
-  - Z2JH_KUBE_VERSION=1.15.6
-  - Z2JH_KUBE_VERSION=1.16.3
-  - Z2JH_KUBE_VERSION=1.16.3 Z2JH_HELM_VERSION=3.0.2
+  global:
+    - K3S_VERSION=v1.18.2+k3s1
+    - HELM_VERSION=v3.2.1
 
 jobs:
-  ## allow experimental setups to fail
-  ##
-  ## ref: https://docs.travis-ci.com/user/customizing-the-build/#rows-that-are-allowed-to-fail
-  ##
-  allow_failures:
-    - name: check docs
   ## don't wait for the jobs that are allowed to fail to report success
   ##
+  ## ref: https://docs.travis-ci.com/user/customizing-the-build/#rows-that-are-allowed-to-fail
   ## ref: https://docs.travis-ci.com/user/customizing-the-build/#fast-finishing
   ##
+  allow_failures:
+    - name: docs:link-check
   fast_finish: true
 
-  ## include additional individual jobs
-  ##
+  ## define individual jobs that can rely on defaults from above
+  ## k3s versions:  https://github.com/rancher/k3s/releases
+  ## helm versions: https://github.com/helm/helm/releases
   include:
-    - stage: lint and validate
+    - stage: test
+      name: chart:k8s-1.18
+
+    - stage: test
+      name: chart:k8s-1.17
+      env: K3S_VERSION=v1.17.5+k3s1 HELM_VERSION=v3.2.1
+
+    - stage: test
+      name: chart:k8s-1.16
+      env: K3S_VERSION=v1.16.9+k3s1 HELM_VERSION=v3.2.1 k3s_disable_command=--no-deploy
```

Comment:
Different versions of kubectl due to different versions of k3s that includes kubectl. We can have a dedicated installstep for latest kubectl to avoid it, ive considered it but didnt care enough either way.

Make a dedicated install of kubectl perhaps?

Reply:
Woops, this was regarding k3s versions indeed. k3s 1.16 introduced a breaking api-change. I'll document that this is the reason.